### DISCUSSION TITLE
Enregistrement de carte Conecs & Swile

### DISCUSSION THREAD

**Discussion by francoistomasi** [PayGreen/documentation#13] [Q&A] [answered]
Bonjour,

Dans notre phase de transition de la v2 à la v3, nous souhaitons reproduire le process d'enregistrement de carte Conecs et Swile que nous avons actuellement.

Cependant avec PayGreenJS, nous n'arrivons pas générer de formulaire pour ces types de cartes.
Seul les `bank_card` fonctionnent pourtant le paramètre `reuse_card_proposal` est bien à `true` pour tous nos moyens de paiements.

```
paygreenjs.init({
    mode: 'instrument',
    modeOptions: {
      authorizedInstrument: true,
    },
    style,
    paymentMethod: 'conecs', // Ou swile
    buyer: 'buy_xxxxxxxxx',
    publicKey: 'pk_xxxxxxxx',
  });
```

Merci d'avance,

**Comment by enzobillisPG** [comment]
Bonjour,
Avez vous une erreur dans la console, ou remontée par l'event ERROR de PGJS ? Avez-vous bien activé les moyens de paiement `conecs` et `swile` depuis votre backoffice ? 

Pour information: `conecs` affichera bien des champs de cartes, cependant pour `swile` cela ouvrira la fenêtre d'Oauth (pas de champs de cartes) 
Enzo

**Comment by francoistomasi** [comment]
Bonjour,

Après quelques tests, il s'avère que le paramètre `authorizedInstrument` doit être à `false` pour que formulaire se génère correctement.

Cependant cela enregistre un instrument avec le paramètre `reuse_allowed: true` mais `authorized: false`.
Ce qui empêche la réutilisation de l'instrument pour un paiement après son enregistrement.

```
{
  "data":{
    "id":"ins_5b8bf5fedb8241dda79c3fa6b57e9f87",
    "object":"instrument",
    "buyer":{
      "id":"buy_aee1863a492d4b329e519a477673f021",
      "object":"buyer",
      "billing_address":null,
      "reference":"PAYGREEN_76",
      "first_name":"xxxxx",
      "last_name":"xxxxx",
      "email":"xxxxxx@xxxxx.xx",
      "phone_number":null,
      "external_id":"cus_xe423cckbrferbric3bhxgt3a4"
    },
    "holder_name":null,
    "platform":"conecs",
    "reuse_allowed":true,
    "expired":false,
    "last_used_at":null,
    "expiry_month":12,
    "expiry_year":2025,
    "scheme":null,
    "last4":"0003",
    "bin":"507597",
    "issuer":"natixis-intertitres",
    "issuer_country":null,
    "daily_balance":5000,
    "authorized":false,
    "authentication":null,
    "billing_typology":null
  },
  "timestamp":"2023-07-05T13:46:10+00:00"
}
```
D'autre part lors de l'enregistrement d'une carte Swile, le paramètre `authorizedInstrument` doit être à `true` pour que la popup soit affichée.
Et là aussi il nous est retourné `reuse_allowed: true` mais `authorized: false`.

```
{
    "id": "ins_1ae5c0c55b624871a0256ddd430d026d",
    "object": "instrument",
    "buyer": {
        "id": "buy_aee1863a492d4b329e519a477673f021",
        "object": "buyer",
        "billing_address": null,
        "reference": "PAYGREEN_76",
        "first_name": "xxxxx",
        "last_name": "xxxxx",
        "email": "xxxxx@xxxxx.xx",
        "phone_number": null,
        "external_id": "cus_xe423cckbrferbric3bhxgt3a4"
    },
    "holder_name": null,
    "platform": "swile",
    "reuse_allowed": true,
    "expired": false,
    "last_used_at": null,
    "expiry_year": null,
    "scheme": null,
    "last4": null,
    "bin": null,
    "issuer": null,
    "issuer_country": null,
    "daily_balance": 2500,
    "authorized": false,
    "authentication": null,
    "billing_typology": null
}
```

Et pour finir, lors de l'enregistrement d'une carte Restoflash, le `buyer` envoyé dans la requête n'est pas retourné dans la réponse.
```
{
  "data": {
    "id": "ins_0d5774e756eb4566ba41165ad378d15d",
    "object": "instrument",
    "buyer": null,
    "holder_name": null,
    "platform": "restoflash",
    "reuse_allowed": true,
    "expired": false,
    "last_used_at": "2023-07-05T09:36:09+00:00",
    "expiry_month": null,
    "expiry_year": null,
    "scheme": null,
    "last4": null,
    "bin": null,
    "issuer": null,
    "issuer_country": null,
    "daily_balance": 2500,
    "authorized": true,
    "authentication": {
      "status": "successed",
      "code": null,
      "authentication_url": null,
      "extra_data": null,
      "payment_config": "pc_8afa2a10bdc8437aa781700775622d3e"
    },
    "billing_typology": null
  },
  "timestamp": "2023-07-05T13:19:58+00:00"
}
```


Merci d'avance pour vos précisions à ce sujet,

**Comment by enzobillisPG** [comment]
Bonjour,

Pour précisions: `authorizedInstrument` doit être passé à `true` lors le moyen de paiement dispose d'un système d'authentification/autorisation. Par exemple, les `bank_card` avec le 3DS, `swile` ou `restoflash` avec l'OAuth. Pour `conecs` il n'y pas de système de ce genre, c'est pourquoi le `authorizedInstrument` est obligatoirement à `false`.

La possibilité de réutilisation d'un instrument ne se base **que** sur `reuse_allowed` qui est complètement indépendant du champ `authorizedInstrument`.

Si vous avez un instrument avec `reuse_allowed` à `true` et que vous avez une erreur lors de sa réutilisable je vous invite à me communiquer son ID ainsi que le message d'erreur retourné. 

Concernant l'instrument Swile qui est envoyé avec `reuse_allowed` à `true` et qui revient à `false` je vais tenter de reproduire et reviens vers vous. 
Pour le buyer qui n'est pas remonté dans l'instrument je remonte votre demande à l'équipe back.

Enzo

**Comment by francoistomasi** [comment]
Bonjour,

Merci pour les précisions sur la propriété `authorizedInstrument`.

Nous avons effectués des jeux de tests sur les différents moyens de paiements.
Tous les moyens de paiements ayant la propriété `reuse_allowed` aboutissent à un paiement validé excepté pour `restoflash` qui demande un `buyer` dans la requête.
Je vous laisse me faire un retour sur la résolution de ce bug.

Merci d'avance,

**Reply by AdrienBr** [reply]
Bonjour @francoistomasi ,
Il n'y a pas de règle spécifique à Restoflash concernant les Buyer. Pourriez-vous nous donner un exemple de Payment Order ou Instrument sur lequel vous rencontrez ce problème ? 

Merci,
Adrien

**Reply by francoistomasi** [reply]
Bonjour @AdrienBr,

Dans le cas de Swile et RestoFlash, le `buyer` n'empêche pas la génération de la popup OAuth.
L'enregistrement de l'instrument se fait bien mais n'est rattaché à aucun `buyer`.

Voici l'instrument retourné lors de l'enregistrement de la carte RestoFlash de test lorsqu'un `buyer` est transmis dans la fonction `paygreenjs.init()`

```
{
  "data": {
    "id": "ins_0d5774e756eb4566ba41165ad378d15d",
    "object": "instrument",
    "buyer": null,
    "holder_name": null,
    "platform": "restoflash",
    "reuse_allowed": true,
    "expired": false,
    "last_used_at": "2023-07-05T09:36:09+00:00",
    "expiry_month": null,
    "expiry_year": null,
    "scheme": null,
    "last4": null,
    "bin": null,
    "issuer": null,
    "issuer_country": null,
    "daily_balance": 2500,
    "authorized": true,
    "authentication": {
      "status": "successed",
      "code": null,
      "authentication_url": null,
      "extra_data": null,
      "payment_config": "pc_8afa2a10bdc8437aa781700775622d3e"
    },
    "billing_typology": null
  },
  "timestamp": "2023-07-05T13:19:58+00:00"
}
```

Et voici le message d'erreur que nous recevons lors de tentative de paiement avec cet instrument
```
{
  "message": "You must provide Buyer informations, or an Instrument linked to a Buyer",
  "timestamp": "2023-07-07T08:10:12+00:00"
}
```

**Reply by AdrienBr** [reply]
@francoistomasi Dans le cas où aucun Buyer n'est rattaché à l'Instrument, il faut obligatoirement donner les infos Buyer à la création du Payment Order.

**Reply by francoistomasi** [reply]
Le problème étant que lorsque l'on récupère les instruments d'un buyer, cet instrument n'est pas affiché.
Il faudrait que dès lors un buyer envoyé dans la fonction `paygreenjs.init()`, ce dernier soit attaché à l'instrument.

Merci pour votre retour

**Reply by enzobillisPG** [reply]
Si je comprends bien, vous avez crée un instrument via le mode instrument de PGJS, en précisant le buyer ID à l'init de PGJS sauf que votre instrument n'as pas été lié au buyer. C'est bien ça ?

**Reply by francoistomasi** [reply]
Exactement, et ce cas se présente uniquement avec la `paymentMethod: 'restoflash'`

**Reply by enzobillisPG** [reply]
Nous allons investiguer et revenir vers vous rapidemment.

**Reply by enzobillisPG** [reply]
Bonjour François,
Après investigations nous avons trouvé deux problèmes:

1. PGJS n'enregistrait plus dans l'instrument le buyer passé en paramètre. Ce qui faisait qu'il était impossible de créer un instrument restoflash avec un buyer.
2. L'API ne retourne pas le buyer dans le call GET instrument.

Pour le premier problème un fix sera déployé avant la fin de semaine et vous permettra donc de créer un instrument restoflash avec buyer. Cependant à cause du deuxième souci (toujours en investigations chez nous) dans l'event INSTRUMENT_READY vous n'aurez pas l'attribut buyer même si ce dernier a bien été link. Je vous tiendrais au courant de l'évolution de ce deuxième souci. 

Bonne journée,
Enzo

**Reply by francoistomasi** [reply]
Bonjour @enzobillisPG,
Merci pour ces informations, nous attendons votre retour quand à la résolution définitive du problème.

**Reply by enzobillisPG** [reply, marked as answer]
Bonjour @francoistomasi,
Je vous informe que le correctif a été déployé en sandbox il y a quelques instants. Il sera déployé en production dans l'après midi. 

Bonne journée,
Enzo
